streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java

```diff
@@ -234,13 +247,15 @@ public V delete(final K key) {
 
     @Override
     public <PS extends Serializer<P>, P> KeyValueIterator<K, V> prefixScan(final P prefix, final PS prefixKeySerializer) {
-
+        Objects.requireNonNull(prefix, "key cannot be null");
```

Comment:
As mentioned by @cadonna the wrapped stores, also check `prefixKeySerializer` for null -- thus might be good to move both check here.

I think we can also remove both checks in `RocksDBStore` and `InMemoryKeyValueStore` -- they seems to be redundant now?

Reply:
>I think we can also remove both checks in RocksDBStore and InMemoryKeyValueStore -- they seem to be redundant now?

they are both different implementations, aren't they?
![image](https://user-images.githubusercontent.com/44339288/115792580-35255c80-a3d3-11eb-99ea-4e5686dd8556.png)
I don't understand how they will be checked if we only leave it in `MeteredKeyValueStore.`

Reply:
![image](https://user-images.githubusercontent.com/44339288/115794170-3441fa00-a3d6-11eb-8fd6-ba9a90af3005.png)
I wrote a quick test to see if it still throws NPE in `InMemoryKeyValueStore` without the check - it did, I am confused, but they are, indeed, redundant.
I'll leave the tests in both `RocksDBStoreTest` and `InMemoryKeyValueStore` bc why not.

Reply:
KafkaStreams runtime _always_ "wraps" any store with a corresponding `MeteredXxxStore` -- those `MeteredXxxsStores` do the transaction from objects to bytes (ie they use the serdes) and also track state store metrics. (Note that stores provided to the runtime always have type `<Bytes, byte[]>` while they are exposed to `Processors` as `<K,V>` types.)

This, when you call `context.stateStore(...)` you always get a `MeteredXxxStore` object -- of course, those details are hidden behind the interface type.

This architecture allows us to unify code and separate concerns. In fact, it also allows us to add/remove more "layers": we can insert a "caching layer" (cf. https://kafka.apache.org/28/documentation/streams/developer-guide/memory-mgmt.html) and a changelogging layer (both are inserted by default).

Reply:
Oooooh, I see.